### ISSUE TITLE
SDDM races with DRM GPU drivers

### ISSUE THREAD

**Issue by theofficialgman** [sddm/sddm#1917]
A similar issue was encountered in GDM and the fix can be found linked in this issue https://gitlab.gnome.org/GNOME/mutter/-/issues/2909

amdgpu driver developer does not want to classify this as an amdgpu driver bug https://gitlab.freedesktop.org/drm/amd/-/issues/3341 as it is independent of the driver in use.

The issue is that on boot the (internal laptop and external displays if connected) are black but backlight is lit. I am able to boot into recovery mode without issue since the graphics drivers are not loaded in that case (only amd framebuffer driver and userspace mesa llvmpipe). Cold boot from OFF seems to be the most common case for this issue and it happens about 50-75% of the time from there.

As seen in the syslog, sddm is starting before amdgpu has fully initialized and this results in sddm starting with a black screen.
https://launchpadlibrarian.net/725987752/preboot.txt

When the sddm starts after amdgpu has fully initialized, it works as intended.
https://launchpadlibrarian.net/725988838/cuttent_boot.txt

This issue has been confirmed to occur on i915 as well with similar frequency.

Alternatives Considered: have systemd `CanGraphical` state change only after initialization has finished https://github.com/systemd/systemd/issues/32509

**Comment by Web-Dev-Codi** [comment]
I am currently running a 7800XT on a B550 MB. I installed Ubuntu Cinnamon 24.04 and got a black screen after the grub menu. The first reboot went fine and stopped working after that. Are there any fixes for this error?

**Comment by theofficialgman** [comment]
Cinnamon traditionally uses LightDM as the display manager, not SDDM which is what this bug report is for. I have not checked LightDM for the same issue but it might happen there too.

The only workarounds currently are restarting the display manager or artificially delaying it's start after the driver has initialized.l (eg: adding a large sleep command to the service that starts the DM)

**Comment by superm1** [comment]
As a "clean" no code workaround, can you add in the systemd unit `ExecStartPre` command something like "udevadm settle"?

**Comment by Web-Dev-Codi** [comment]
Thanks. This opens up more I can look into for a possible fix. I installed a pre-release 2 weeks ago and it worked flawlessly. This error only occurred with the official release.   I hope I can solve this soon.

**Comment by Vogtinator** [comment]
> Alternatives Considered: have systemd CanGraphical state change only after initialization has finished https://github.com/systemd/systemd/issues/32509

Yeah, that's the way to go. This is all just a giant hack, so it should be worked around in a single place which is systemd.

> As a "clean" no code workaround, can you add in the systemd unit ExecStartPre command something like "udevadm settle"?

FWICT there's also a `systemd-udev-settle.service` which is meant for legacy services but fits perfectly here. Please try https://github.com/sddm/sddm/pull/1924.

**Comment by Web-Dev-Codi** [comment]
Can someone break this down in lehman's terms? I have no idea what this means and how to find a workaround.

**Comment by Web-Dev-Codi** [comment]
Excuse my new noobishness. I enjoy the cinnamon flavor for my web development and would love to use 24 soon

**Comment by Vogtinator** [comment]
> Can someone break this down in lehman's terms? I have no idea what this means and how to find a workaround.

If you use ubuntu cinnamon, this is the wrong place anyway, see https://github.com/sddm/sddm/issues/1917#issuecomment-2081511941.

**Comment by theofficialgman** [comment]
> > Alternatives Considered: have systemd CanGraphical state change only after initialization has finished [systemd/systemd#32509](https://github.com/systemd/systemd/issues/32509)
> 
> Yeah, that's the way to go. This is all just a giant hack, so it should be worked around in a single place which is systemd.
> 
> > As a "clean" no code workaround, can you add in the systemd unit ExecStartPre command something like "udevadm settle"?
> 
> FWICT there's also a `systemd-udev-settle.service` which is meant for legacy services but fits perfectly here. Please try #1924.

I won't be able to reliably report back the results of adding that I think. My racecondition is too small.
It varies from AMDGPU finishing loading ~1 second before sddm starts to finishing loading ~1 second after sddm starts.
Some days it wants to start everytime, othertimes nearly never. I imagine calling the settle service with its sleeps just waiting for the daemon to start will be enough to push sddm starting later enough on its own, regardless of the checks it does.

Also systemd-udev-settle.service and udevadm settle (they are the same thing) are deprecated and you aren't supposed to use them https://github.com/systemd/systemd/blob/main/src/udev/udevadm-settle.c

**Comment by superm1** [comment]
I see the pull request as a no code workaround until there is a better solution like a code change to systemd or to sddm to do the settling actions on PCI VGA class devices.

**Comment by shoffmeister** [comment]
In the past I had regular problems with boot-stuck SDDM on an Intel+NVIDIA notebook for a while, see https://bugs.kde.org/show_bug.cgi?id=478616 - could those be related?

This problem manifested in log output such as
```
kwin_core: Failed to open /dev/dri/card0 device (Did not receive a reply.
```
and black screens.

FWIW, I am now on Fedora 40 / KDE 6 proper, and I do not recollect any such misbehaviour in recent times. I am unable to trace back this "magically disappeared" to any specific change: 
* all the drivers changed (NVIDIA, Intel), together with the kernel
* removed any/all NVIDIA and Intel driver overrides
* cleaned up some pre-F40 SDDM overrides (towards Wayland)
* removed some environment variables on KDE which set DRM device (card) order

**Comment by superm1** [comment]
The discussion upstream on dri-devel has mostly settled upon userspace greeters need to support hot-unplug.
https://lore.kernel.org/dri-devel/ZkyZCmMU86nUV4TO@phenom.ffwll.local/

**Comment by darksylinc** [comment]
This is an old problem https://github.com/sddm/sddm/pull/1324

But it's back on SDDM. It's not surprising, since the solution back then was to just wait a little longer and try again, which is not a proper solution.

**Comment by Duvio** [comment]
I recently discovered this in my logfiles:
```
$ sudo systemctl status systemd-udev-settle.service
● systemd-udev-settle.service - Wait for udev To Complete Device Initialization
     Loaded: loaded (/usr/lib/systemd/system/systemd-udev-settle.service; static)
     Active: active (exited) since Thu 2024-09-19 10:28:27 CEST; 6min ago
       Docs: man:systemd-udev-settle.service(8)
   Main PID: 568 (code=exited, status=0/SUCCESS)
        CPU: 2ms

Sep 19 10:28:22 nbnrzdev005 systemd[1]: Starting systemd-udev-settle.service - Wait for udev To Complete Device Initialization...
Sep 19 10:28:22 nbnrzdev005 udevadm[568]: systemd-udev-settle.service is deprecated. Please fix sddm.service not to pull it in.
Sep 19 10:28:27 nbnrzdev005 systemd[1]: Finished systemd-udev-settle.service - Wait for udev To Complete Device Initialization.
```

In total, this delays my boot-up-time by 4.5 seconds.

**Comment by sfornengo** [comment]
It seems linked to my black screen at boot on kubuntu 24.04.1 with nvidia driver.
So what is the workaround ?

**Comment by Conan-Kudo** [comment]
This issue is coming up a lot more on Fedora with amdgpu based systems since Fedora 43. See [rhbz#2404966](https://bugzilla.redhat.com/show_bug.cgi?id=2404966).

Downstream discussion seems to indicate SDDM needs an equivalent of this logic to fix the issue: https://gitlab.gnome.org/GNOME/gdm/-/commit/895f765aa8cc5a9dd2901be65bcd638b8aa7c577

**Comment by superm1** [comment]
FYI that change in gnome was changed later on for kernel changes. Here is the adaptation for newer kernels.

https://gitlab.gnome.org/GNOME/gdm/-/commit/200cf5452e03f846de6d7001592b42d138ecf326

**Comment by Thulium-Drake** [comment]
Hi there all!

I seem to have a similar issue with SDDM _after_ logging into a sway session, but only when multiple screens have been connected to my laptop. I was able to find a workaround, but first the deetz:

- Framework 13 with AMD Ryzen 7 7840U w/ Radeon  780M Graphics
- Latest BIOS (this was a thing with earlier versions) - 03.16
- Fedora 43 with the following packages:
```
[root@archon ~]# rpm -qa | grep -E 'sddm|sway'
swayidle-1.8.0-8.fc43.x86_64
swaybg-1.2.1-4.fc43.x86_64
swaylock-1.8.3-1.fc43.x86_64
sway-1.11-2.fc43.x86_64
sway-config-fedora-0.4.3-2.fc43.noarch
sddm-0.21.0-10.fc43.x86_64
sddm-wayland-sway-0.4.3-2.fc43.noarch
sway-systemd-0.4.1-3.fc43.noarch
swayimg-4.5-2.fc43.x86_64
```

To re-iterate, this _only_ happens when the laptop is hooked up to a dock with 2 additional monitors

When logging in, the screen is blank, no cursor, no session, nothing. When logging in on the laptop via SSH and collecting logs I see the following in `journalctl --user -r`:

```
sway[4006]: 00:00:00.790 [ERROR] [sway/server.c:247] Unable to create backend
sway[4006]: 00:00:00.786 [ERROR] [wlr] [backend/backend.c:420] Failed to open any DRM device
sway[4006]: 00:00:00.786 [ERROR] [wlr] [backend/backend.c:245] Found 0 GPUs, cannot create backend
sway[4006]: 00:00:00.777 [ERROR] [wlr] [backend/session/session.c:331] Failed to open device: '/dev/dri/card1': Resource temporarily unavailable
sway[4006]: 00:00:00.777 [ERROR] [wlr] [libseat] [libseat/backend/logind.c:124] Could not take device: Device or resource busy
```

The main reason I suspect `sddm` here is due to the following:

- `sddm` is supposed to start `sway` and it works perfectly fine on my laptop without additional monitors
- Removing `sddm` from the mix does _not_ show similar issues
- I added a 'sleep 1` in '/usr/bin/start-sway` as a means to give `sddm` more time to free up the GPU and the issue was worked around succesfully.

So the question remains, this sounds to me like a race condition, but who's to blame (`sddm` too slow releasing GPU resources, or `sway` too fast in wanting to acquire them)? :-)

And is this even related to the issue at hand (if not, please, feel free to split it off or let me know and I will make a new issue)?

**Comment by superm1** [comment]
> Hi there all!
> 
> I seem to have a similar issue with SDDM _after_ logging into a sway session, but only when multiple screens have been connected to my laptop. I was able to find a workaround, but first the deetz:
> 
> - Framework 13 with AMD Ryzen 7 7840U w/ Radeon  780M Graphics
> - Latest BIOS (this was a thing with earlier versions) - 03.16
> - Fedora 43 with the following packages:
> ```
> [root@archon ~]# rpm -qa | grep -E 'sddm|sway'
> swayidle-1.8.0-8.fc43.x86_64
> swaybg-1.2.1-4.fc43.x86_64
> swaylock-1.8.3-1.fc43.x86_64
> sway-1.11-2.fc43.x86_64
> sway-config-fedora-0.4.3-2.fc43.noarch
> sddm-0.21.0-10.fc43.x86_64
> sddm-wayland-sway-0.4.3-2.fc43.noarch
> sway-systemd-0.4.1-3.fc43.noarch
> swayimg-4.5-2.fc43.x86_64
> ```
> 
> To re-iterate, this _only_ happens when the laptop is hooked up to a dock with 2 additional monitors
> 
> When logging in, the screen is blank, no cursor, no session, nothing. When logging in on the laptop via SSH and collecting logs I see the following in `journalctl --user -r`:
> 
> ```
> sway[4006]: 00:00:00.790 [ERROR] [sway/server.c:247] Unable to create backend
> sway[4006]: 00:00:00.786 [ERROR] [wlr] [backend/backend.c:420] Failed to open any DRM device
> sway[4006]: 00:00:00.786 [ERROR] [wlr] [backend/backend.c:245] Found 0 GPUs, cannot create backend
> sway[4006]: 00:00:00.777 [ERROR] [wlr] [backend/session/session.c:331] Failed to open device: '/dev/dri/card1': Resource temporarily unavailable
> sway[4006]: 00:00:00.777 [ERROR] [wlr] [libseat] [libseat/backend/logind.c:124] Could not take device: Device or resource busy
> ```
> 
> The main reason I suspect `sddm` here is due to the following:
> 
> - `sddm` is supposed to start `sway` and it works perfectly fine on my laptop without additional monitors
> - Removing `sddm` from the mix does _not_ show similar issues
> - I added a 'sleep 1` in '/usr/bin/start-sway` as a means to give `sddm` more time to free up the GPU and the issue was worked around succesfully.
> 
> So the question remains, this sounds to me like a race condition, but who's to blame (`sddm` too slow releasing GPU resources, or `sway` too fast in wanting to acquire them)? :-)
> 
> And is this even related to the issue at hand (if not, please, feel free to split it off or let me know and I will make a new issue)?

Maybe try using Wayland for SDDM instead?

**Comment by Conan-Kudo** [comment]
This is using Wayland. SDDM on the Fedora Sway spin is configured with the Sway compositor.

**Comment by Thulium-Drake** [comment]
@superm1 I don't think my `sddm` is running in X, but in wayland instead. This is `ps aux | grep sddm` without any user logged in:

```
root        2391  0.0  0.0 159876 23540 ?        Ssl  20:05   0:00 /usr/bin/sddm
root        2443  0.0  0.0  79316 21180 ?        S    20:05   0:00 /usr/libexec/sddm-helper --socket /tmp/sddm-auth-751d5f9a-3677-4f9f-8e99-465b1c108759 --id 2 --start /usr/bin/sddm-greeter-qt6 --socket /tmp/sddm--Hlbbzn --theme /usr/share/sddm/themes/03-sway-fedora --user sddm --display-server /usr/libexec/sddm-compositor-sway --greeter
sddm        2518  0.0  0.0  23840 14884 ?        Ss   20:05   0:00 /usr/lib/systemd/systemd --user
sddm        2624  0.0  0.0  36536  3408 ?        S    20:06   0:00 (sd-pam)
sddm        2718  0.0  0.0  64328 13944 tty1     Ss+  20:06   0:00 /usr/libexec/sddm-helper-start-wayland /usr/libexec/sddm-compositor-sway /usr/bin/sddm-greeter-qt6 --socket /tmp/sddm--Hlbbzn --theme /usr/share/sddm/themes/03-sway-fedora
sddm        2723  0.5  0.2 816988 119164 tty1    Sl+  20:06   0:00 /usr/bin/sway --config /etc/sway/sddm-greeter.config
sddm        3433  2.5  0.4 1987800 262944 tty1   Sl+  20:06   0:04 /usr/bin/sddm-greeter-qt6 --socket /tmp/sddm--Hlbbzn --theme /usr/share/sddm/themes/03-sway-fedora
sddm        3442  0.0  0.0   5444  3552 ?        Ss   20:06   0:00 swayidle -w timeout 300 swaymsg "output * power off" resume swaymsg "output * power on"
sddm        3443  0.0  0.0  51196 32048 ?        Ss   20:06   0:00 /usr/bin/python3 /usr/libexec/sway-systemd/locale1-xkb-config
sddm        3447  0.0  0.0   8956  4400 ?        Ss   20:06   0:00 /usr/bin/dbus-broker-launch --scope user
sddm        3448  0.0  0.0   5076  3140 ?        S    20:06   0:00 dbus-broker --log 11 --controller 10 --machine-id 6b305dd3aaa14545bf017f53403ad3e1 --max-bytes 100000000000000 --max-fds 25000000000000 --max-matches 5000000000
```

**Comment by LorenzoBianconi** [comment]
I can reproduce the issue on my desktop running Fedora43 + KDE + wayland
```
Linux lore-desk 6.17.9-300.fc43.x86_64 #1 SMP PREEMPT_DYNAMIC Mon Nov 24 23:31:27 UTC 2025 x86_64 GNU/Linux
sddm-0.21.0-10.fc43.x86_64
AMD Ryzen 9 9950X3D 16-Core Processor
VGA compatible controller: Advanced Micro Devices, Inc. [AMD/ATI] Granite Ridge [Radeon Graphics] (rev c9) (prog-if 00 [VGA controller])
        Subsystem: ASUSTeK Computer Inc. Device 8877
        ...
        Kernel driver in use: amdgpu
        Kernel modules: amdgpu

```
X11 does not trigger any issue

Adding the suggested patch below fixes the issue for me
```
--- ../sddm.service	2025-12-03 13:04:12.373397892 +0100
+++ /usr/lib/systemd/system/sddm.service	2025-12-03 13:05:31.184269695 +0100
@@ -2,13 +2,16 @@
 Description=Simple Desktop Display Manager
 Documentation=man:sddm(1) man:sddm.conf(5)
 Conflicts=getty@tty1.service
-After=systemd-user-sessions.service getty@tty1.service plymouth-quit.service systemd-logind.service
+After=systemd-user-sessions.service getty@tty1.service plymouth-quit.service systemd-logind.service systemd-udev-settle.service
 PartOf=graphical.target
 StartLimitIntervalSec=30
 StartLimitBurst=2
+# Wait for driver loading to be complete so that graphics work
+# (https://github.com/sddm/sddm/issues/1917)
+Wants=systemd-udev-settle.service
 
 [Service]
 ExecStart=/usr/bin/sddm
```

**Comment by mcgi5sr2** [comment]
issue present on Asahi Alarm as well.

```
[libseat/backend/logind.c:121] Could not take device: No such device
no drm device found
fatal: failed to create compositor backend
```

OS: Arch Linux ARM aarch64
Kernel: Linux 6.18.15-1-1-ARCH

**Comment by Serafadam** [comment]
I'm not sure if this is the exact issue but I stumbled upon it during my search, my issue was as follows:
Black screen with only mouse cursor visible on Fedora 43  / Hyprland
I tried fixes above but unfortunately had no luck, what seems to have fixed it was adding
```
add_drivers+=" nvidia nvidia_modeset nvidia_uvm nvidia_drm "
```
to 
```
/etc/dracut.conf.d/nvidia.conf
```
Then 
```
sudo dracut --force
```

**Comment by mkurz** [comment]
Related in plasma-login-manager:
- https://invent.kde.org/plasma/plasma-login-manager/-/merge_requests/130

**Comment by freevatar** [comment]
Same issue here:

`Fedora 44`

**Comment by mvdan** [comment]
I started running into this only when upgrading from Fedora 43 to 44 on my Framework 13 with a Ryzen HX 370. I raised an issue with Fedora Sway here, to see if they can assist with any solution or workaround on their end. https://gitlab.com/fedora/sigs/sway/SIG/-/work_items/45

**Comment by GregDMeyer** [comment]
> I started running into this only when upgrading from Fedora 43 to 44 on my Framework 13 with a Ryzen HX 370. I raised an issue with Fedora Sway here, to see if they can assist with any solution or workaround on their end. https://gitlab.com/fedora/sigs/sway/SIG/-/work_items/45

I also only started running into this issue after upgrading to Fedora 44. Was ok on Fedora 43. Also on a framework 13 here, with 7840U processor.
